File: vane-regions/src/main/java/org/oddlama/vane/regions/region/EnvironmentSetting.java
package org.oddlama.vane.regions.region;

import org.oddlama.vane.regions.Regions;

public enum EnvironmentSetting {
    // Spawning
    ANIMALS(true),
    MONSTERS(false),

    // Hazards
    EXPLOSIONS(false),
    FIRE(false),
    PVP(true),

    // Environment
    TRAMPLE(false),
    VINE_GROWTH(false);

    private boolean def;

    private EnvironmentSetting(final boolean def) {
        this.def = def;
    }

    public boolean defaultValue() {
        return def;
    }

    public boolean hasOverride() {
        return getOverride() != 0;
    }

    public int getOverride() {
        return Regions.environmentOverrides.getOverride(this);
    }
}

File: vane-regions/src/main/kotlin/org/oddlama/vane/regions/region/EnvironmentSetting.kt
package org.oddlama.vane.regions.region

import org.oddlama.vane.regions.Regions

/**
 * Region-level environment behaviors that can be configured or overridden.
 */
enum class EnvironmentSetting(private val def: Boolean) {
    // Spawning
    ANIMALS(true),
    MONSTERS(false),

    // Hazards
    EXPLOSIONS(false),
    FIRE(false),
    PVP(true),

    // Environment
    TRAMPLE(false),
    VINE_GROWTH(false);

    /**
     * Returns the default value when no group or global override is set.
     */
    fun defaultValue(): Boolean = def

    /**
     * Returns whether a global override currently applies to this setting.
     */
    fun hasOverride(): Boolean = override != 0

    /**
     * Global override value from `Regions.environmentOverrides`.
     */
    val override: Int
        get() = Regions.environmentOverrides?.getOverride(this) ?: 0
}
